checker/src/main/java/org/checkerframework/checker/nullness/NullnessTransfer.java

```diff
@@ -345,33 +356,54 @@ public TransferResult<NullnessValue, NullnessStore> visitThrow(
         return result;
     }
 
-    /*
-     * Provided that m is of a type that implements interface java.util.Map:
+    /**
+     * {@inheritDoc}
+     *
+     * <p>When the conservativeArgumentNullnessAfterInvocation flag is turned off, the receiver and
+     * arguments that are passed to non-null parameters in a method call or constructor invocation
+     * are unsoundly assumed to be non-null after the invocation.
+     *
+     * <p>When the flag is turned on, the analysis is more conservative by checking the method is
+     * SideEffectFree or the receiver is unassignable. Only if either one of the two is true, is the
+     * receiver made non-null. Similar logic is applied to the arguments of the invocation.
+     *
+     * <p>Provided that m is of a type that implements interface java.util.Map:
+     *
      * <ul>
-     * <li>Given a call m.get(k), if k is @KeyFor("m") and m's value type is @NonNull,
-     *     then the result is @NonNull in the thenStore and elseStore of the transfer result.
+     *   <li>Given a call m.get(k), if k is @KeyFor("m") and m's value type is @NonNull, then the
+     *       result is @NonNull in the thenStore and elseStore of the transfer result.
      * </ul>
      */
     @Override
     public TransferResult<NullnessValue, NullnessStore> visitMethodInvocation(
             MethodInvocationNode n, TransferInput<NullnessValue, NullnessStore> in) {
         TransferResult<NullnessValue, NullnessStore> result = super.visitMethodInvocation(n, in);
 
-        // Make receiver non-null.
+        MethodInvocationTree tree = n.getTree();
+        ExecutableElement method = TreeUtils.elementFromUse(tree);
+
+        boolean isMethodSideEffectFree = PurityUtils.isSideEffectFree(atypeFactory, method);
         Node receiver = n.getTarget().getReceiver();
-        makeNonNull(result, receiver);
+        if (nonNullAssumptionAfterInvocation
+                || isMethodSideEffectFree
+                || JavaExpression.fromNode(receiver).isUnassignableByOtherCode()) {
```

Comment:
If you look at the test failure, you see this (partial) diff:

````
/__w/1/s/checker/bin/javac -AnoJreVersionCheck -processor org.checkerframework.checker.nullness.NullnessChecker -Acfgviz=org.checkerframework.dataflow.cfg.visualize.StringCFGVisualizer -AassumeKeyFor Issue5174.java > Out.txt 2>&1 || true
diff -u Issue5174.out Out.txt
--- Issue5174.out	2022-09-02 14:29:49.934145186 +0000
+++ Out.txt	2022-09-02 14:35:55.113600746 +0000
@@ -101,7 +101,6 @@
 15:
 Before: NullnessStore#12(
   f > NV{, S, poly nn/n = f/f}
-  this > NV{@UnknownInitialization @NonNull, Issue5174Super, poly nn/n = f/f}
   Issue5174Super<S>.sf > NV{@Initialized @NonNull, Object, poly nn/n = f/f}
   initialized fields = [sf]
   invariant fields = [sf]
@@ -131,7 +130,6 @@
 12:
 Before: NullnessStore#14(
   f > NV{, S, poly nn/n = f/f}
-  this > NV{@UnknownInitialization @NonNull, Issue5174Super, poly nn/n = f/f}
   Issue5174Super<S>.sf > NV{@Initialized @NonNull, Object, poly nn/n = f/f}
````

That is, `this` references are no longer in the store.
That is odd, because `isUnassignableByOtherCode()` should always return `true`:
https://github.com/eisop/checker-framework/blob/e8175116002f699e0ebe146edb69bb086681c325/dataflow/src/main/java/org/checkerframework/dataflow/expression/ThisReference.java#L41

Do you understand why this is failing?